### DISCUSSION TITLE
Achieve multiple Class/Type membership? (instance of)

### DISCUSSION THREAD

**Discussion by lapidus** [geldata/gel#5631] [Q&A] [answered]
EdgeDB supports multiple inheritance via Object Types. For example the type `Museum` could inherit properties from `Place` and `Institution` via extension.

But sometimes it is also helpful for an instance to have **multiple type memberships**. 

For example the instance "Louvre Museum" could be seen as both a `Tourist Attraction` with certain properties and a `Museum` with other properties.

See this example from Wikidata:

https://www.wikidata.org/wiki/Q19675

<img width="939" alt="image" src="https://github.com/edgedb/edgedb/assets/66601/590b7935-a8df-4e28-8fe7-198d89a0ce52">

Has there been any discussions/work on supporting such a scenario?

**Comment by tailhook** [comment, marked as answer]
There are multiple ways you can model that. The inheritance approach might provoke a combinatorial explosion of the types:
```
type TouristAttraction {
  season: TouristSeason,
}
type Museum {
  working_hours: str,
}
type TouristAttraction_Museum extends TouristAttraction, Museum; # lots of combinations here
```

So it's better approached with links. There are still few ways of doing that. Straightforward is probably:
```
type Base {
  name: str;
}
type TouristAttraction extending Base {
  season: TouristSeason;
}
type Museum extending Base {
  working_hours: str;
}
type Item {
  multi memberships: Base;
  # optional shortcuts to make it easier to access
  link as_museum := .memberships[IS Museum];
  link as_tourist_attraction := .memberships[IS TouristAttraction];
  # typecheck shortcuts
  property is_museum := exists(.as_museum);
  property is_tourist_attraction := exists(.as_tourist_attraction);
  property all_types := .memberships.__type__.name;
}
```
The alternative that links from the other side, might be helpful to ensure that there is only one Museum/TouristAttraction membership:
```
type Base {
  name: str;
  member: Item;
}
type TouristAttraction extending Base {
  season: TouristSeason;
  constraint exclusive on (.member);  # each type has to have its own constraint
}
type Museum extending Base {
  working_hours: str;
  constraint exclusive on (.member);  # each type has to have its own constraint
}
type Item {
  # shortcuts to make it easier to access
  link as_museum := .<member[IS Museum];
  link as_tourist_attraction := .<member[IS TouristAttraction];
}
```

In both cases, you can easily select item with all their specific membership properties:
```
SELECT Item { 
  name,
  as_museum { working_hours },
  as_tourist_abstraction { season },
} FILTER .name = <str>$name
```
You can also flatten the result:
```
SELECT Item { 
  name,
  working_hours := as_museum.working_hours,
  season := as_tourist_abstraction.season,
} FILTER .name = <str>$name
```
And you don't have to have `as_` computables, you can do the same within the query too:
```
SELECT Item { 
  name,
  working_hours := .<member[IS Museum].working_hours,
  season := .<member[IS TouristAttraction].season,
} FILTER .name = <str>$name
```

**Reply by lapidus** [reply]
Wow, that's some bad-ass modelling ;) Thank you for the detailed examples!

Could you help me unpack what the below statement does? Not yet super familiar with all the schema goodies.

link as_museum := .memberships[IS Museum];

Thank you!

**Reply by tailhook** [reply]
We call this "computable". This helps you keep some expressions in the schema that you could otherwise write in every query:

I.e. this query:
```
SELECT Item {
  name,
  as_museum := .memberships[IS Museum] {
    working_hours
  },
}
```
Is equivalent to:
```
SELECT Item {
  name,
  as_museum {
    working_hours
  },
}
```
If you have `as_museums` computable defined.

More details:
```
link as_museum := .memberships   [IS Museum];
^^^^              ^^^^^^^^^^^^   ^^^^^^^^^^^
 |                     |              \-- filter link by type `Museum`
 |                     |                            
 |                     \-- Read link `memberships`
 |              
 \--- merely marks computable that returns a link (i.e. set of objcects rather than scalars)

```

**Reply by lapidus** [reply]
Thank you, much appreciated! 🙏
